ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/BlockPropagationManager.java

```diff
@@ -533,26 +539,59 @@ private CompletableFuture<Block> scheduleGetBlockFromPeers(
       final Optional<EthPeer> maybePreferredPeer,
       final long blockNumber,
       final Optional<Hash> maybeBlockHash) {
-    final RetryingGetBlockFromPeersTask getBlockTask =
-        RetryingGetBlockFromPeersTask.create(
-            protocolSchedule,
-            ethContext,
-            config,
-            metricsSystem,
-            Math.max(1, ethContext.getEthPeers().peerCount()),
-            maybeBlockHash,
-            blockNumber);
-    maybePreferredPeer.ifPresent(getBlockTask::assignPeer);
-
-    var future =
-        ethContext
-            .getScheduler()
-            .scheduleSyncWorkerTask(getBlockTask::run)
-            .thenCompose(r -> importOrSavePendingBlock(r.getResult(), r.getPeer().nodeId()));
-
-    ethContext.getScheduler().failAfterTimeout(future, getBlockTimeoutMillis);
-
-    return future;
+
+    return ethContext
+        .getScheduler()
+        .scheduleServiceTask(
+            () -> {
+              GetHeadersFromPeerTask headersFromPeerTask =
+                  new GetHeadersFromPeerTask(
+                      maybeBlockHash.orElse(null),
+                      blockNumber,
+                      1,
+                      0,
+                      Direction.FORWARD,
+                      Math.max(1, ethContext.getEthPeers().peerCount()),
+                      protocolSchedule);
+              if (maybePreferredPeer.isPresent()) {
+                PeerTaskExecutorResult<List<BlockHeader>> executorResult =
+                    ethContext
+                        .getPeerTaskExecutor()
+                        .executeAgainstPeer(headersFromPeerTask, maybePreferredPeer.get());
+                if (executorResult.result().isPresent()
+                    && executorResult.responseCode() != PeerTaskExecutorResponseCode.SUCCESS) {
+                  return CompletableFuture.completedFuture(
```

Comment:
This logic looks wrong. If we are not successful shouldn't we return a failedFuture?